### ISSUE TITLE
The memory is full when training with a large amount of data

### ISSUE THREAD

**Issue by hanmy1021** [MaartenGr/BERTopic#522]
I used 385000 abstracts to fit a bertopic model. If I set calculate_probabilities=False, it will be 7 hours to fit a model and there is no problem with memory. But if I set calculate_probabilities=True cause I want to get some probabilities and embeddings, it will be more than 12 hours to fit a model and it will run out of 32g of memory. How could I improve my method? Here is my code:

`df = pd.read_csv('/root/abstract.csv', engine='python', sep=',')
docs = df['Abstract'].to_list()  #385000 abstracts
model = BERTopic(nr_topics=100, calculate_probabilities=True)
topics, probabilities = model.fit_transform(docs)
model.save('/root/test_model')
`

**Comment by hanmy1021** [comment]
`df = pd.read_csv('/root/abstract.csv', engine='python', sep=',')`
`docs = df['Abstract'].to_list() #385000 `
`abstracts model = BERTopic(nr_topics=100, calculate_probabilities=True) `
`topics, probabilities = model.fit_transform(docs) `
`model.save('/root/test_model')`

**Comment by hanmy1021** [comment]
It has been running for 26 hours. I'm wondering if it's deadlocked. 
I can provide some running details of CPU, GPU and memory.
When it started running, the GPU usage was 100%, which lasted about 5-10 minutes. After that the GPU usage was 0%, the CPU usage was 800%(eight core) and 15GB of memory(about 50%) has been used, which lasted about a few hours. 
Now it is still running, the GPU usage is still 0%, the CPU usage is 100%(it seems to be only one core running) and the 32GB of memory(almost 100%) has been used.

**Comment by MaartenGr** [comment]
Could you share BERTopic's version, the version of its dependencies (i.e., HDBSCAN, Numpy, Numba, UMAP, scikit-learn), and the python version you are using?

To start off, it might be helpful to set `verbose=True` to get an idea of where the model seems to take the most time. Moreover, creating the embeddings beforehand allows you to speed up the re-training of the model. You can find a bit more about that [here](https://maartengr.github.io/BERTopic/getting_started/embeddings/embeddings.html). 

Coming back to the issue you are having. Setting `calculate_probabilities=True` will significantly slow down the application as HDBSCAN needs to calculate the soft-clustering in order to extract the probabilities. If your HDBSCAN install is not version v0.8.28, then it would be worthwhile to upgrade it as it solves quite a number of issues with large datasets.

**Comment by hanmy1021** [comment]
> Could you share BERTopic's version, the version of its dependencies (i.e., HDBSCAN, Numpy, Numba, UMAP, scikit-learn), and the python version you are using?
> 
> To start off, it might be helpful to set `verbose=True` to get an idea of where the model seems to take the most time. Moreover, creating the embeddings beforehand allows you to speed up the re-training of the model. You can find a bit more about that [here](https://maartengr.github.io/BERTopic/getting_started/embeddings/embeddings.html).
> 
> Coming back to the issue you are having. Setting `calculate_probabilities=True` will significantly slow down the application as HDBSCAN needs to calculate the soft-clustering in order to extract the probabilities. If your HDBSCAN install is not version v0.8.28, then it would be worthwhile to upgrade it as it solves quite a number of issues with large datasets.

Thank you a lot for your reply! I will share you some version:
bertopic:0.9.4
hdbscan:0.8.28
numpy:1.21.3
numba:0.55.1
umap:0.5.3
sklearn:1.0.1
python:3.8.12
It seems the version of HDBSCAN is the latest. I will try the embedding models.
And I wonder if the dataset is too large cause the memory will always run out?(The dataset csv is 713MB with 385000 abstracts)

**Comment by MaartenGr** [comment]
Ah, in that case, it is most likely due to how HDBSCAN handles calculating soft probabilities. It is an expensive procedure to do and will, as a result, create a *n* x *m* matrix, where *n* is the number of documents and *m* is the number of topics. Depending on your settings, the resulting matrix can therefore be quite large especially since it is in a dense format. Unfortunately, there currently is no way of speeding this process up whilst still using `calculate_probabilities=True` unless you use a smaller dataset or decrease the number of topics by increasing `min_topic_size`.

**Comment by hanmy1021** [comment]
> Ah, in that case, it is most likely due to how HDBSCAN handles calculating soft probabilities. It is an expensive procedure to do and will, as a result, create a _n_ x _m_ matrix, where _n_ is the number of documents and _m_ is the number of topics. Depending on your settings, the resulting matrix can therefore be quite large especially since it is in a dense format. Unfortunately, there currently is no way of speeding this process up whilst still using `calculate_probabilities=True` unless you use a smaller dataset or decrease the number of topics by increasing `min_topic_size`.

Ah, it explains the matter, thank you again! There is one more question, could you tell me is there a way to avoid classifying the topic of the document into -1 when I set calculate_probabilities=False? Before I set calculate_probabilities=True cause I wanted to reclassify documents with a topic of - 1 into another topic with a second probability. Now it seem difficult to achieve. So I wander if there is another way to do it.

**Comment by MaartenGr** [comment]
> There is one more question, could you tell me is there a way to avoid classifying the topic of the document into -1 when I set calculate_probabilities=False? Before I set calculate_probabilities=True cause I wanted to reclassify documents with a topic of - 1 into another topic with a second probability. Now it seem difficult to achieve. So I wander if there is another way to do it.

Good question, if `calculate_probabilities=False` is set and cannot be True due to computational and/or memory issues, then that definitely creates a problem when it comes to removing outliers. 

There are some things you can do with HDBSCAN as mentioned [here](https://maartengr.github.io/BERTopic/faq.html#how-do-i-reduce-topic-outliers) that might reduce the number of outliers. However, it has worked for some but definitely not for all. 

Instead, and this is rather good timing, you can leverage some of the new features in BERTopic v0.10.0. It was released a few days ago an allows you to us k-Means instead of HDBSCAN. By using k-Means, no outliers will be generated but it might increase the number of stop words in the resulting topic representations which fortunately can be removed with the CountVectorizer. 

To start off, install the new version with `pip install --upgrade bertopic`. Then, simply run the following to use k-Means instead:

```python
from bertopic import BERTopic
from sklearn.cluster import KMeans
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model = CountVectorizer(stop_words="english")
cluster_model = KMeans(n_clusters=50)
topic_model = BERTopic(hdbscan_model=cluster_model, vectorizer_model=vectorizer_model)
```

You can find a bit more about this new feature in the changelog [here](https://maartengr.github.io/BERTopic/changelog.html) or following the documentation [here](https://maartengr.github.io/BERTopic/getting_started/dim_reduction/dim_reduction.html) and [here](https://maartengr.github.io/BERTopic/getting_started/clustering/clustering.html).

Hopefully, this helps a bit!

**Comment by hanmy1021** [comment]
> > There is one more question, could you tell me is there a way to avoid classifying the topic of the document into -1 when I set calculate_probabilities=False? Before I set calculate_probabilities=True cause I wanted to reclassify documents with a topic of - 1 into another topic with a second probability. Now it seem difficult to achieve. So I wander if there is another way to do it.
> 
> Good question, if `calculate_probabilities=False` is set and cannot be True due to computational and/or memory issues, then that definitely creates a problem when it comes to removing outliers.
> 
> There are some things you can do with HDBSCAN as mentioned [here](https://maartengr.github.io/BERTopic/faq.html#how-do-i-reduce-topic-outliers) that might reduce the number of outliers. However, it has worked for some but definitely not for all.
> 
> Instead, and this is rather good timing, you can leverage some of the new features in BERTopic v0.10.0. It was released a few days ago an allows you to us k-Means instead of HDBSCAN. By using k-Means, no outliers will be generated but it might increase the number of stop words in the resulting topic representations which fortunately can be removed with the CountVectorizer.
> 
> To start off, install the new version with `pip install --upgrade bertopic`. Then, simply run the following to use k-Means instead:
> 
> ```python
> from bertopic import BERTopic
> from sklearn.cluster import KMeans
> from sklearn.feature_extraction.text import CountVectorizer
> 
> vectorizer_model = CountVectorizer(stop_words="english")
> cluster_model = KMeans(n_clusters=50)
> topic_model = BERTopic(hdbscan_model=cluster_model, vectorizer_model=vectorizer_model)
> ```
> 
> You can find a bit more about this new feature in the changelog [here](https://maartengr.github.io/BERTopic/changelog.html) or following the documentation [here](https://maartengr.github.io/BERTopic/getting_started/dim_reduction/dim_reduction.html) and [here](https://maartengr.github.io/BERTopic/getting_started/clustering/clustering.html).
> 
> Hopefully, this helps a bit!

Thank you a lot! The kmeans completely solved my problem. And I would like to ask how much is the accuracy difference between kmeans and HDBSCAN? The topic model I got using kmeans also looks quite accurate.

**Comment by MaartenGr** [comment]
> And I would like to ask how much is the accuracy difference between kmeans and HDBSCAN?

That is a difficult question and actually depends on your definition of `accuracy`. Is that topic coherence, topic diversity, quality of clusters (RAND index), etc? 

With the subjectivity that is topic modeling, I would advise focusing on metrics that are relevant to your use case. In this case, if your results look quite accurate and you feel that they represent the documents well, they should be accurate enough.

**Comment by MaartenGr** [comment]
Due to inactivity, this issue will be closed. Feel free to ping me if you want to re-open the issue!
